src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "lib/chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import "lib/yield-utils-v2/contracts/token/IERC20.sol";
+import "lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+
+contract Vault is Ownable {
+
+    ///@dev Vault records collateral deposits of each user
+    mapping (address => uint) public deposits; 
+
+    ///@dev Vault records debt holdings of each user
+    mapping (address => uint) public debts;  
+
+
+    ///@notice ERC20 interface specifying token contract functions
+    ///@dev For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
+    IERC20 public immutable collateral;    
+    IERC20 public immutable debt;  
+
+    ///@dev Asset Pricefeed interface from Chainlink
+    AggregatorV3Interface public immutable priceFeed;   
+    
+    ///@dev Emitted on deposit()
+    ///@param collateralAsset The address of the collateral asset
+    ///@param user The address of the user calling deposit()
+    ///@param collateralAmount The amount of collateral asset deposited
+    event Deposit(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on borrow()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling borrow()
+    ///@param debtAmount The amount of debt asset borrowed
+    event Borrow(address indexed debtAsset, address indexed user, uint debtAmount); 
+
+    ///@dev Emitted on repay()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling repay()
+    ///@param debtAmount The amount of debt asset being repaid
+    event Repay(address indexed debtAsset, address indexed user, uint debtAmount);
+
+    ///@dev Emitted on withdraw()
+    ///@param collateralAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param collateralAmount The amount of collateral asset withdrawn
+    event Withdraw(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on liquidation()
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@dev To set collateralization level of debt engine
+    ///@notice To allow for margin call at defined level, preventing Vault from bearing risk in fast moving markets
+    uint public immutable bufferDenominator; 
```

Comment:
this feels like something that we would might want to change later, but then u have to use a storage var which kinda sucks for gas usage.  if u do change this to a storage var i would use a ray instead of storing both numerator and denominator. in fact i would probably do that here anyways.  even a ray might be too much, how many digits of precision do u really need?  🤔

Reply:
in my imagining, these levels are usually set in some integer value btw 70-80%, at least for retail fx brokers - so not much precision would be required. 

as for changing it, I'm not sure. brokers rarely modify margin call levels, but perhaps my analogy itself might be off-tangent, you'll have to correct me 😕

Reply:
if not much precision is required, then this is even more of an argument to use a fp number.  as for changing it, i see u what u mean about margin calls.  at yield we have different collateralization rates depending on the base and collateral -- and sometimes we make mistakes or change our mind about the collateralization ratio (also sometimes external factors could arise that might make us want to change)  so thats where i was coming from.

for this exercise u can leave it as immutable, but i would suggest a fp number instead of 2 integers, numerator/denominator